pkg/internal/approver/validation/validator.go

```diff
@@ -0,0 +1,90 @@
+package validation
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/ext"
+)
+
+const (
+	varSelf    = "self"
+	varRequest = "cr"
+)
+
+// Validator knows how to validate CSR attribute values in CertificateRequests
+// against CEL expressions declared in CertificateRequestPolicy.
+// A Validator must be compiled before it can validate. A compiled Validator
+// is stateless, thread-safe, and cacheable.
+type Validator interface {
+	// Compile compiles the Validator CEL expression into a runnable program
+	// that can perform validations. Any compilation errors will be returned.
+	Compile() error
```

Comment:
@erikgb I think we can still improve this interface. I see two approaches:
- Compile() could return a new struct that has a Validate function.
- Validate() could lazily call Compile() when necessary & we can make the Compile() function private

Currently you have to remember whether Compile was called already, making it harder to use this interface.

Please let me know what you think.

Reply:
For now, we could remove `Compile` from the interface, as the cache compiles the validator (if new validator) before returning it to the caller. Your first suggestion will force us to put names on TWO interfaces, and we all know that naming is the hardest part of programming. 😉 The webhook will typically only compile the expression, to verify it's valid, and never invoke `Validate`. So I don't think your second suggestion is a good one - mainly because you suggest hiding away `Compile`. I generally think we should separate `Compile` from `Validate`,  mainly because the error handling must be different for those two operations.

Reply:
Removed `Compile` from `Validator` interface.